### PR TITLE
Preemption callback fix

### PR DESCRIPTION
**Pull request by jbieniusiewi**
# What does this PR do ?

Preemption callback `nemo/utils/callbacks/preemption.py` terminates a training process with `sys.exit`.
Hence other PTL callbacks/hooks, that are normally executed at the end of a training, are not run.

In particular async checkpointing finalisation `AsyncFinalizerCallback` might be not completed. 

This PR replaces manual checkpoint saving and `sys.exit` with `trainer.should_stop=True` for a graceful shutdown.
(same mechanism is implemented in a new preemption callback `nemo/lightning/pytorch/callbacks/preemption.py`)

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/utils/callbacks/preemption.py [hunk 1]
```diff
@@ -91,10 +91,8 @@ def on_train_batch_end(self, trainer, pl_module, outputs, batch, batch_idx: int)
             # a regular local variable
             interrupted = self.interrupted
             if interrupted:
-                logging.info("Received SIGTERM, saving checkpoint and exiting")
-                monitor_candidates = self.checkpoint_callback._monitor_candidates(trainer)
-                self.checkpoint_callback._save_last_checkpoint(trainer, monitor_candidates)
-                sys.exit(0)
+                logging.info("Preemption detected, signaling trainer to stop")
+                trainer.should_stop = True
```
Comment (athitten) [inline nemo/utils/callbacks/preemption.py:95]: @jbieniusiewi how is the last checkpoint saved in this case ?

Comment (jbieniusiewi) [reply nemo/utils/callbacks/preemption.py:95]: In my experiments, when should_stop=True, it consistently: runs the evaluation loop and saves the checkpoint before terminating.  I'm not sure if there are some PTL flags that control this behaviour.

@cuichenx I've noticed that you've added a new preemption callback that stops the training with `trainer.should_stop=True`. Could you please share your experience with this? Specifically, does the checkpoint get saved when `trainer.should_stop` is set to True?

Comment (jbieniusiewi) [reply nemo/utils/callbacks/preemption.py:95]: I'm closing this PR, due to the possible problems with the last checkpoint saving.

Comment (cuichenx) [reply nemo/utils/callbacks/preemption.py:95]: Hi @jbieniusiewi I don't think I added it.. which one are you referring to?

Comment (jbieniusiewi) [reply nemo/utils/callbacks/preemption.py:95]: Hi @cuichenx 

By the "new preemption callback" I meant: `nemo/lightning/pytorch/callbacks/preemption.py`

[Which uses "trainer.should_stop=True" to stop the training.](https://github.com/NVIDIA/NeMo/blob/68d4dba331fdafbb33a640f4e3ed20fc660864a7/nemo/lightning/pytorch/callbacks/preemption.py#L65)

But it seems it has been recently fixed [on "r2.0.1" branch](https://github.com/NVIDIA/NeMo/blob/da0b0b7cae62358f2023851fd04ac90d123bc159/nemo/lightning/pytorch/callbacks/preemption.py#L72)

And the updated version uses this construct to save the checkpoint and stop the training:

```
  monitor_candidates = trainer.checkpoint_callback._monitor_candidates(trainer)
  trainer.checkpoint_callback._save_last_checkpoint(trainer, monitor_candidates)
  sys.exit(0)
```

So I infer that using "trainer.should_stop=True" is not correct.

Comment (athitten) [reply nemo/utils/callbacks/preemption.py:95]: @jbieniusiewi , @hemildesai added it in `nemo/lightning/pytorch/callbacks/preemption.py`. He should be able to provide comments for that.

**Pull request closed without merging**